src/ui/events.js

```diff
@@ -0,0 +1,576 @@
+ * @property {Coordinate} [coord] The coordinate of the tile if the event has a `dataType` of `source` and
+ * the event is related to loading of a tile.
+ */
+export type MapDataEvent = {
+    type: string,
+    dataType: string
+};
+
+export type MapContextEvent = {
+    type: 'webglcontextlost' | 'webglcontextrestored',
+    originalEvent: WebGLContextEvent
+}
+
+export type MapEvent =
+    /**
+     * Fired when a pointing device (usually a mouse) is pressed within the map.
+     *
+     * @event mousedown
+     * @memberof Map
+     * @instance
+     * @property {MapMouseEvent} data
+     * @see [Highlight features within a bounding box](https://www.mapbox.com/mapbox-gl-js/example/using-box-queryrenderedfeatures/)
+     * @see [Create a draggable point](https://www.mapbox.com/mapbox-gl-js/example/drag-a-point/)
+     */
+    | 'mousedown'
+
+    /**
+     * Fired when a pointing device (usually a mouse) is released within the map.
+     *
+     * @event mouseup
+     * @memberof Map
+     * @instance
+     * @property {MapMouseEvent} data
+     * @see [Highlight features within a bounding box](https://www.mapbox.com/mapbox-gl-js/example/using-box-queryrenderedfeatures/)
+     * @see [Create a draggable point](https://www.mapbox.com/mapbox-gl-js/example/drag-a-point/)
+     */
+    | 'mouseup'
+
+    /**
+     * Fired when a pointing device (usually a mouse) is moved within the map.
+     *
+     * @event mousemove
+     * @memberof Map
+     * @instance
+     * @property {MapMouseEvent} data
+     * @see [Get coordinates of the mouse pointer](https://www.mapbox.com/mapbox-gl-js/example/mouse-position/)
+     * @see [Highlight features under the mouse pointer](https://www.mapbox.com/mapbox-gl-js/example/hover-styles/)
+     * @see [Display a popup on hover](https://www.mapbox.com/mapbox-gl-js/example/popup-on-hover/)
+     */
+    | 'mousemove'
+
+    /**
+     * Fired when a pointing device (usually a mouse) is pressed and released at the same point on the map.
+     *
+     * @event click
+     * @memberof Map
+     * @instance
+     * @property {MapMouseEvent} data
+     * @see [Measure distances](https://www.mapbox.com/mapbox-gl-js/example/measure/)
+     * @see [Center the map on a clicked symbol](https://www.mapbox.com/mapbox-gl-js/example/center-on-symbol/)
+     */
+    | 'click'
+
+    /**
+     * Fired when a pointing device (usually a mouse) is clicked twice at the same point on the map.
+     *
+     * @event dblclick
+     * @memberof Map
+     * @instance
+     * @property {MapMouseEvent} data
+     */
+    | 'dblclick'
+
+    | 'mouseenter'
+    | 'mouseleave'
+    | 'mouseover'
+
+    /**
+     * Fired when a point device (usually a mouse) leaves the map's canvas.
+     *
+     * @event mouseout
+     * @memberof Map
+     * @instance
+     * @property {MapMouseEvent} data
+     * @see [Highlight features under the mouse pointer](https://www.mapbox.com/mapbox-gl-js/example/hover-styles/)
```

Comment:
hmm this example actually uses the `mouseleave` event, not `mouseout`.  Are we not defining the events on L165-167 because they're self explanatory or for some other reason?